File: Days 21 - 30/Day 22/GetOriginalSentenceFromString.cpp
#include <iostream>
#include <set>
#include <string>
#include <vector>

std::vector<std::string> GetOriginalSentence(const std::string& sentence, const std::set<std::string>& dictionary)
{
	std::string searchString = sentence;
	std::vector<std::string> words;

	while (searchString != "")
	{
		bool wordFound = false;

		for (const auto& word : dictionary)
		{
			if (searchString.find(word) == 0)
			{
				words.push_back(word);

				searchString = searchString.substr(word.length());
				wordFound = true;
			}
		}

		if (!wordFound)
		{
			return std::vector<std::string> { };
		}
	}

	return words;
}

template <typename T>
void PrintVector(const std::vector<T>& values) noexcept
{
	if (values.empty())
	{
		std::cout << "Empty vector...";
	}

	for (const auto& value : values)
	{
		std::cout << value << " ";
	}

	std::cout << "\n";
}

int main(int argc, const char* argv[])
{
	std::set<std::string> dictionary { "quick", "brown", "the", "fox" };
	std::string sentence = "thequickbrownfox";

	PrintVector(GetOriginalSentence(sentence, dictionary));

	std::set<std::string> secondDictionary { "bed", "bath", "bedbath", "and", "beyond" };
	std::string secondSentence = "bedbathandbeyond";

	PrintVector(GetOriginalSentence(secondSentence, secondDictionary));

	std::set<std::string> thirdDictionary { "this", "will", "return", "an", "empty", "vector" };
	std::string thirdSentence = "nodictionarywordsinhere";

	PrintVector(GetOriginalSentence(thirdSentence, thirdDictionary));

	std::cin.get();

	return 0;
}

File: Days 021 - 030/Day 22/GetOriginalSentenceFromString.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Day22
	{
		private static int Main(string[] args)
		{
			SortedSet<string> dictionary = new SortedSet<string> { "quick", "brown", "the", "fox" };
			string sentence = "thequickbrownfox";
			PrintList(GetOriginalSentence(sentence, dictionary));

			dictionary = new SortedSet<string> { "bed", "bath", "bedbath", "and", "beyond" };
			sentence = "bedbathandbeyond";
			PrintList(GetOriginalSentence(sentence, dictionary));

			dictionary = new SortedSet<string> { "this", "will", "return", "an", "empty", "list" };
			sentence = "nodictionarywordsinhere";
			PrintList(GetOriginalSentence(sentence, dictionary));

			Console.ReadLine();

			return 0;
		}

		private static List<string> GetOriginalSentence(string sentence, SortedSet<string> dictionary)
		{
			List<string> words = new List<string>();
			string searchString = sentence;

			while (!string.IsNullOrEmpty(searchString))
			{
				bool wordFound = false;

				foreach (string word in dictionary)
				{
					if (searchString.StartsWith(word))
					{
						words.Add(word);

						searchString = searchString.Substring(word.Length);
						wordFound = true;
					}
				}

				if (!wordFound)
				{
					return new List<string>();
				}
			}

			return words;
		}

		private static void PrintList<T>(List<T> list)
		{
			if (list.Count == 0)
			{
				Console.WriteLine("List is empty...");

				return;
			}

			foreach (T value in list)
			{
				Console.Write($"{value} ");
			}

			Console.WriteLine();
		}
	}
}